bn-apps/cordapp-updates-distribution/README.md

```diff
@@ -0,0 +1,270 @@
+
+## Session Filters
+
+Corda-based transports allow developers to implement their custom `SessionFilter`s that can be used to reject any unintended download requests. `SessionFilter` is a simple interface that, if implemented, is invoked against every incoming download request. `SessionFilter` returns a boolean that indicates whether the request should be let through or not.
+
+```kotlin
+interface SessionFilter {
+    @Suspendable
+    fun isSessionAllowed(session : FlowSession, flowLogic : FlowLogic<*>) : Boolean
+}
+```
+For example a session filter that would allow only the Business Network traffic through can be implemented in the following way:
+
+```kotlin
+class BusinessNetworkSessionFilter : SessionFilter {
+    @Suspendable
+    override fun isSessionAllowed(session : FlowSession, flowLogic : FlowLogic<*>) = flowLogic.subFlow(GetMembershipsFlow())[session.counterparty] != null
+}
+``` 
+
+## Transport modes
+
+The following transports are supported:
+* **corda-flows** allows to transfer data over Corda Flows. It can be used from *inside a Corda node* only. 
+* **corda-rpc** allows to transfer data over Corda RPC. It reuses the same flows as `corda-flows` transport, but invokes them via RPC instead.  
+* **corda-auto** is an automatic switch between `corda-rpc` and `corda-flows` transports. The underlying transport is chosen based on the value of `corda-updates.mode` custom session property, that is set in the runtime based on the invocation context. The main purpose for this mode - is to allow Cordapp Distribution Service to reuse the same configuration file, regardless of whether it was invoked from inside or outside of a Corda node. 
+
+Corda-based transports expect a remote repository URL to be specified in the format of `transport-name:x500Name`. For example `corda-auto:O=BNO,L=New York,C=US` (just imagine that you specify Corda X500 name instead of HTTP hostname).
+
+**corda-auto** is the recommended way of using Corda-based transports.
+
+### Asynchronous invocations
+
+Maven Resolver over Corda-based transports should always be invoked from a *non-flow thread* when used *inside Corda Flows*. This is because, under the hood Corda-based transports start a separate flow (which starts in a different from the calling thread) for data transfer to prevent Maven Resolver internals from being checkpointed as they are not `@Suspendable`. As Corda OS flows engine is single-threaded, invoking the transports synchronously would result into a deadlock, where the calling flow would be indefinitely waiting for the transport flow to finish while the transport flow would be indefinitely waiting for the calling flow to finish to be able to start.
```

Comment:
Yes, kind of. What would have been great - if we could enable compatibility behaviours depending on the platform version in runtime. Something like https://developer.android.com/training/basics/supporting-devices/platforms, otherwise the app would work on the enterprise only.

Reply:
Maybe we should just bring checkpoint skipping into open source instead.

Reply:
That would be great